### ISSUE TITLE
Add support for specifying outbound source IP address (i.e. for multi-homed servers)

### ISSUE THREAD

**Issue by cipriancraciun** [albertito/chasquid#22]
In case of multi-homed servers (having multiple public interfaces, with multiple default routes but using routing policies) it would be useful to have an option to specify the local IP address that should be used to connect from (i.e. `bind(local_address)` then `connect(remote_address)`).

This should be configurable through the configuration file, for example: `smtp_local_address = "9.9.9.9"`.

In fact, by default this IP address should be initialized with the IP from `smtp_address`, as most likely the IP used to accept inbound SMTP connections should also be used for outbound SMTP connections.  (And in case that is missing, just use `0.0.0.0` as it currently is.)

----

Looking at the source code, I assume that one could just change:
https://github.com/albertito/chasquid/blob/master/internal/courier/smtp.go#L128
~~~~
	conn, err := net.DialTimeout("tcp", mx+":"+*smtpPort, smtpDialTimeout)
~~~~
with
~~~~
        localAddress = ...
        remoteAddress, err := net.ResolveTCPAddr("tcp", mx+":"+*smtpPort)
        ...
	conn, err := net.DialTcp("tcp", localAddress, remoteAddress)
        ...
~~~~

However that loses the `smtpDialTimeout` functionality.

Or perhaps using the `net.Dialer` type, although that might be somewhat more complex.

**Comment by albertito** [comment]
Sorry I didn't reply before, I thought I had :(

Thanks for the suggestion!

I need to read up/refresh my knowledge on route policies a bit, think about the implications of this more carefully. It's not clear to me this is worth the complexity, but it definitely could be.

If there's more information that would help prioritize this higher, (e.g. more folks running into this, no possible workarounds, etc.) please let me know.

Thank you!

**Comment by cipriancraciun** [comment]
> I need to read up/refresh my knowledge on route policies a bit, think about the implications of this more carefully. It's not clear to me this is worth the complexity, but it definitely could be.


In essence `chasquid` shouldn't care at all about the routing;  using (Linux) routing policies was just an example how one could achieve this (and doesn't directly involve `chasquid`), there are also other techniques (such as a router downstream switching the provider based on the source IP).

Just `bind`-ing on a given IP (as opposed to the default `0.0.0.0` should be enough) to implement this feature.


> If there's more information that would help prioritize this higher, (e.g. more folks running into this, no possible workarounds, etc.) please let me know.


I'm not sure many would use this feature (as many servers have one IP and one default route).  Regarding work-arounds there are, but they aren't as straight-forward as having `chasquid` bind on the desired IP.

**Comment by ThinkChaos** [comment]
I think this should work to implement it with a context:
```go
d := net.Dialer{
	// LocalAddr = ...
}

ctx, cancel := context.WithTimeout(context.Background(), smtpDialTimeout)
defer cancel()

conn, err := d.DialContext(ctx, "tcp", mx+":"+*smtpPort)
if err != nil {
	return a.tr.Errorf("Could not dial: %v", err), false
}
defer conn.Close()
```

**Comment by Satvik200** [comment]
I'd like to contribute a solution for this issue. Here's my proposed approach:

### Changes
1. Add `smtp_local_address` configuration option
2. Use `net.Dialer` to bind outbound connections:
   ```go
   dialer := &net.Dialer{
       Timeout: smtpDialTimeout,
   }
   if localAddr != nil {
       dialer.LocalAddr = localAddr
   }
   conn, err := dialer.Dial("tcp", net.JoinHostPort(mx, *smtpPort))

**Comment by cipriancraciun** [comment]
> There are also other outgoing connections, like the ones used for DNS lookups, or MTA-STS. Are those also expected to be covered by this option, or just SMTP? 

Indeed, there might be many places where one would (potentially) want to source the connections from a different IP, and indeed one could extend this feature even further to have per target domain source IP.  This could quickly become a mess...

However, when it comes to email delivery (due to the world we currently live in), I think there are two major types of traffic (mainly based on how ISPs treat them):
* auxiliary -- DNS, MTA-STS, and potentially any other kind of TCP/UDP traffic that is auxiliar to SMTP;  this is just "normal traffic" for ISPs, and they don't touch it, or even care about it;
* port 25 (and alternatives) -- i.e. the dreaded SMTP delivery port which every ISP wants you to jump through hoops to allow you to use it (both inbound and especially outbound, if they even allow it);

Thus, I think just adding an option to configure the source of outbound SMTP traffic would suffice to solve the majority of those use-cases where such a feature would be needed (which as granted previously might just be a tiny fraction of normal use-cases).

Perhaps, just to cover all ground, and if easily implemented, also source at least the MTA-STS traffic from the same IP, and if possible the DNS resolution.

----

> Another possibility I am considering is to document how to set this up with cgroups+iptables+iproute2, with examples, to make it easier to set up for the folks who need it.

But this just covers Linux (and leaves out the BSDs).

And not even the entirety of Linux deployments, because:
* there are (although very few) distributions that don't default to `systemd` (mainly Alpine Linux);
* `iptables` used to be the only Linux firewall system, but now we also have `nftables` which uses a completely different syntax;
* and even if all of the above is documented, unfortunately many Linux distributions provide wrappers to handle `iptables` / `nftables`, like `firewalld` (and I believe NetworkManager and `systemd-networkd` also have support for configuring it), which unfortunately believe they are the owners of the entire `iptables` / `nftables` ruleset, and any change you make outside of this "firewall manager", just randomly disappears the next time the firewall manager is tickled;

Thus, although documentation might help, it would be mostly out-of-date or not applicable to the whole combinatorial explosion of solutions out there.

**Comment by albertito** [comment]
> Thus, I think just adding an option to configure the source of outbound SMTP traffic would suffice to solve the majority of those use-cases where such a feature would be needed (which as granted previously might just be a tiny fraction of normal use-cases).

This is fair, and as long as it's clear in the docs that it only covers outgoing connections on port 25, it's doable. But I am worried that most folks who want that feature, will also want it to be applied to all outgoing traffic. And...

> Perhaps, just to cover all ground, and if easily implemented, also source at least the MTA-STS traffic from the same IP, and if possible the DNS resolution.

... doing it for all traffic is significant complexity, that has to be supported forever, at arguably the wrong layer.

I understand it's a useful feature, and I know it can be implemented for port 25 as it is today. I'm being very transparent about my reasoning around the tradeoffs of adding it, and why I haven't done it yet.

> But this just covers Linux (and leaves out the BSDs).

Most modern systems support source routing in one way or another. In Linux you can use netns or source routing. In FreeBSD you can use setfib. In OpenBSD you can use rdomain.

It's just each is different, and the configuration is non-trivial.

That's why one of the options I am considering is  giving examples of how to do this at the OS level in common scenarios. I understand it's impossible to cover all scenarios and that wouldn't be the goal anyway.

That may be less convenient than a built-in setting to the users that need to do it, but it may be a reasonable tradeoff.

**Comment by cipriancraciun** [comment]
In the end, I think the decision about this feature request boils down to a simple question:  do you intend to add extensive customization features to `chasquid`?  (Including different source IP for SMTP, DNS, STS, etc.)

If not, then perhaps closing this issue is the best solution.  (And I'm OK with closing this issue.)

----

(as a sidenote)

For example, just after my last reply in June, I've hit another minor snag with my deployment:  not being able to specify a different DNS resolver than the system's default (from `/etc/resolv.conf`).

On the same server where I have `chasquid` running, I also have a custom DNS resolver that implements (poorly) split horizon, and it failed to resolve TXT records for my own domain, thus all SPAM email that was sent to me (and made to look as coming from my own domain), didn't properly apply SPF policy (because the TXT record for my own domain didn't resolve properly).

----

Indeed, putting `chasquid` into a container, or namespace, would solve both my issues:
* I can easily control the source IP through policy routing;
* I can bind-mount a different `resolv.conf` inside the namespace;

However, this complicates the setup quite a lot.

I liked `chasquid` because it's trivial to deploy:  just a small set of self-contained binaries, just one server process, just a few configuration files.  If I need to resort to namespaces, it'll be perhaps easier to switch to a Docker / Podman solution, and at that point, all the simplicity of `chasquid` is gone.
